miranda/decode/_decoder.py

```diff
@@ -227,16 +242,51 @@ def _decode_time_info(
                 if field == "timedelta":
                     return pd.to_timedelta(time_dictionary[potential_times[0]])
                 return time_dictionary[potential_times[0]]
-        elif data:
+
+        if file and data:
+            file_parts = Path(file).name.split("_")
+            potential_times = [
+                segment for segment in file_parts if segment in time_dictionary.keys()
+            ]
             potential_time = data["frequency"]
             if potential_time == "":
                 time_units = data["time"].units
                 potential_time = time_units.split()[0]
-            if field == "timedelta":
-                if potential_time in ["fx", "fixed"]:
-                    return pd.NaT
-                return pd.to_timedelta(time_dictionary[potential_time])
-            return time_dictionary[potential_time]
+
+            if potential_time in potential_times:
+                return time_dictionary[potential_time]
+
+            else:
+                logging.warning(
+                    f"Frequency from metadata not found in filename: `{Path(file).name}`: "
+                    "Performing more rigorous frequency checks."
+                )
+                if Path(file).is_file() and Path(file).suffix in [".nc", ".nc4"]:
+                    engine = "netcdf4"
+                elif Path(file).is_dir() and Path(file).suffix == ".zarr":
+                    engine = "zarr"
+                else:
+                    raise DecoderError(file)
+                _, found_freq = get_time_frequency(
+                    xarray.open_dataset(
+                        file,
+                        engine=engine,
+                        drop_variables="time_bnds",
+                    ).time
+                )
+
+                if found_freq in potential_times:
+                    return found_freq
+                elif found_freq == "month":
+                    for f in ["Amon", "Omon", "monC", "monthly", "months", "mon"]:
+                        if f in potential_times:
+                            return time_dictionary[f]
+                else:
+                    logging.warning(
+                        "Time frequency found in dataset on analysis was not found in filename. "
+                        f"Basing fields on `{found_freq}`."
```

Comment:
Maybe add information in the warning to say if `found_freq` is the same or different from the frequency in the metadata.